packages/frontend/app/routes/payments._index.tsx

```diff
@@ -0,0 +1,177 @@
+import { json, type LoaderArgs } from '@remix-run/node'
+import { useLoaderData, useNavigate, useSearchParams } from '@remix-run/react'
+import { Badge, PageHeader } from '~/components'
+import { PopoverFilter } from '~/components/Filters'
+import { Button, Table } from '~/components/ui'
+import { listPayments } from '~/lib/api/payments.server'
+import { paymentsSearchParams } from '~/lib/validate.server'
+import { PaymentType } from '~/generated/graphql'
+import type { CombinedPaymentState } from '~/shared/utils'
+import { capitalize, badgeColorByState } from '~/shared/utils'
+
+export const loader = async ({ request }: LoaderArgs) => {
+  const url = new URL(request.url)
+  const searchParams = Object.fromEntries(url.searchParams.entries())
+
+  const result = paymentsSearchParams.safeParse(
+    searchParams.type
+      ? { ...searchParams, type: searchParams.type.split(',') }
+      : searchParams
+  )
+
+  if (!result.success) {
+    throw json(null, { status: 400, statusText: 'Invalid result.' })
+  }
+
+  const { type, ...pagination } = result.data
+
+  const payments = await listPayments({
+    ...pagination,
+    ...(type ? { filter: { type: { in: type } } } : {})
+  })
+
+  let previousPageUrl = '',
+    nextPageUrl = ''
+
+  if (payments.pageInfo.hasPreviousPage) {
+    previousPageUrl = `/payments?before=${payments.pageInfo.startCursor}`
+  }
+
+  if (payments.pageInfo.hasNextPage) {
+    nextPageUrl = `/payments?after=${payments.pageInfo.endCursor}`
+  }
+
+  return json({ payments, previousPageUrl, nextPageUrl, type })
+}
+
+export default function PaymentsPage() {
+  const { payments, previousPageUrl, nextPageUrl, type } =
+    useLoaderData<typeof loader>()
+  const setSearchParams = useSearchParams()[1]
+
+  const navigate = useNavigate()
+
+  function setTypeFilterParams(selectedType: PaymentType): void {
+    const selected = type
+
+    if (!selected.includes(selectedType)) {
+      selected.push(selectedType)
+      setSearchParams(new URLSearchParams({ type: selected.join(',') }))
+      return
+    }
+
+    setSearchParams(() => {
+      const newParams = selected.filter((t) => t !== selectedType).join(',')
+      if (newParams.length === 0) {
+        return ''
+      }
+
+      return new URLSearchParams({
+        type: newParams
+      })
+    })
+  }
+
+  const paymentSubpathByType: {
+    [key in PaymentType]: string
+  } = {
+    [PaymentType.Incoming]: 'incoming',
+    [PaymentType.Outgoing]: 'outgoing'
+  }
```

Comment:
This can also be moved outside the component.

Reply:
moved all of these